cmd/dump/src/lib.rs

```diff
@@ -48,34 +48,895 @@
+    emulate_dumper: bool,
+
     dumpfile: Option<String>,
 }
 
+//
+// When using the dump agent, we create our own ersatz Core
+//
+#[derive(Default)]
+struct AgentCore {
+    flash_contents: Vec<u8>,
+    flash_regions: BTreeMap<u32, (u32, usize)>,
+    ram_regions: BTreeMap<u32, Vec<u8>>,
+    registers: HashMap<ARMRegister, u32>,
+}
+
+impl AgentCore {
+    fn new(hubris: &HubrisArchive) -> Result<AgentCore> {
+        //
+        // We want to read in the "final.elf" from our archive and use that
+        // to populate our flash memory to dump.
+        //
+        let cursor = Cursor::new(hubris.archive());
+        let mut archive = zip::ZipArchive::new(cursor)?;
+        let mut file = archive
+            .by_name("img/final.elf")
+            .map_err(|e| anyhow!("failed to find final.elf: {}", e))?;
+
+        let mut flash_contents = Vec::new();
+        file.read_to_end(&mut flash_contents)?;
+
+        let elf = Elf::parse(&flash_contents).map_err(|e| {
+            anyhow!("failed to parse final.elf as an ELF file: {}", e)
+        })?;
+
+        let mut flash_regions = BTreeMap::new();
+
+        for shdr in elf.section_headers.iter() {
+            if shdr.sh_type != goblin::elf::section_header::SHT_PROGBITS {
+                continue;
+            }
+
+            flash_regions.insert(
+                shdr.sh_addr as u32,
+                (shdr.sh_size as u32, shdr.sh_offset as usize),
+            );
+        }
+
+        Ok(Self { flash_contents, flash_regions, ..Default::default() })
+    }
+
+    fn add_ram_region(&mut self, addr: u32, contents: Vec<u8>) {
+        self.ram_regions.insert(addr, contents);
+    }
+
+    fn add_register(&mut self, reg: ARMRegister, val: u32) {
+        self.registers.insert(reg, val);
+    }
+
+    fn read_flash(&self, addr: u32, data: &mut [u8]) -> Result<()> {
+        if let Some((&base, &(size, offset))) =
+            self.flash_regions.range(..=addr).rev().next()
+        {
+            if base > addr || base + size <= addr {
+                //
+                // It's not here, and we have already tried RAM...
+                //
+                bail!("address 0x{:08x} not found", addr);
+            }
+
+            let start = (addr - base) as usize;
+            let roffs = offset + start;
+
+            if start + data.len() <= size as usize {
+                //
+                // This flash region wholly contains our desired region; copy
+                // it and leave.
+                //
+                data.copy_from_slice(
+                    &self.flash_contents[roffs..roffs + data.len()],
+                );
+
+                return Ok(());
+            }
+
+            let len = (size as usize) - start;
+            data[..len]
+                .copy_from_slice(&self.flash_contents[roffs..roffs + len]);
+
+            self.read_flash(addr + len as u32, &mut data[len..])
+        } else {
+            bail!("address 0x{:08x} not found", addr);
+        }
+    }
```

Comment:
Nit: this may be more clear as an explicit loop, instead of recursing

```suggestion
    fn read_flash(&self, mut addr: u32, mut data: &mut [u8]) -> Result<()> {
        while !data.is_empty() {
            let (&base, &(size, offset)) =
                self.flash_regions.range(..=addr).last().ok_or_else(|| {
                    anyhow!("address 0x{:08x} not found", addr)
                })?;

            if !(base..base + size).contains(&addr) {
                //
                // It's not here, and we have already tried RAM...
                //
                bail!("address 0x{:08x} not found", addr);
            }

            let start = (addr - base) as usize;
            let roffs = offset + start;

            let amount = (size as usize - start).min(data.len());
            data[0..amount]
                .copy_from_slice(&self.flash_contents[roffs..roffs + amount]);
            addr += amount as u32;
            data = &mut data[amount..];
        }
        Ok(())
    }
```

Reply:
You also prevent a possible stack overflow that way. I typically like recursion, but stack overflows make me sad.

Reply:
I guess if its only a few sections, probably not really an issue.

Reply:
Yeah, dunno on that one. To me, it's clearer as recursion -- but I may chalk this one up to personal preference?

Reply:
Yup, I don't feel too strongly about it.